File: scripts/update-navigation.py
#!/usr/bin/env python3
"""
Script to update navigation in all HTML files with the new navigation from navigation.html
"""

import os
import re
import glob
from pathlib import Path

# Get the base directory
BASE_DIR = Path(__file__).parent.parent

# Read the new navigation
NAV_FILE = BASE_DIR / "navigation.html"
with open(NAV_FILE, 'r', encoding='utf-8') as f:
    NEW_NAV = f.read().strip()

# Pattern to match navigation section
# Matches from <!-- Navigation --> or <nav class="navigation" to </nav>
NAV_PATTERN = re.compile(
    r'(<!--\s*Navigation\s*-->)?\s*<nav\s+class=["\']navigation["\'].*?</nav>',
    re.DOTALL | re.IGNORECASE
)

# Alternative pattern for pages that might have different structure
NAV_PATTERN_ALT = re.compile(
    r'<!--\s*Navigation\s*Placeholder\s*-->.*?<script.*?fetch.*?navigation\.html.*?</script>',
    re.DOTALL | re.IGNORECASE
)

def find_navigation_section(content):
    """Find the navigation section in content"""
    # Try main pattern first
    match = NAV_PATTERN.search(content)
    if match:
        return match
    
    # Try alternative pattern
    match = NAV_PATTERN_ALT.search(content)
    if match:
        return match
    
    return None

def update_file(file_path):
    """Update navigation in a single file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Skip if it's the navigation.html file itself
        if file_path.name == 'navigation.html':
            return False, "Skipped (navigation.html itself)"
        
        # Check if file already has the new navigation (check for mobile-nav-container)
        if 'mobile-nav-container' in content and 'id="mobile-nav-container"' in content:
            return False, "Already has new navigation"
        
        # Find navigation section
        nav_match = find_navigation_section(content)
        
        if not nav_match:
            return False, "No navigation found"
        
        # Replace the navigation
        new_content = content[:nav_match.start()] + NEW_NAV + content[nav_match.end():]
        
        # Write back
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(new_content)
        
        return True, "Updated"
    
    except Exception as e:
        return False, f"Error: {str(e)}"

def main():
    """Main function to update all HTML files"""
    # Find all HTML files
    html_files = []
    for pattern in ['**/*.html', '**/*.htm', '**/*.shtml']:
        html_files.extend(glob.glob(str(BASE_DIR / pattern), recursive=True))
    
    # Filter out node_modules and other directories
    html_files = [f for f in html_files if 'node_modules' not in f and 'metallography.org' not in f]
    
    print(f"Found {len(html_files)} HTML files to check")
    print(f"New navigation from: {NAV_FILE}")
    print("-" * 80)
    
    updated = 0
    skipped = 0
    errors = 0
    
    for file_path in sorted(html_files):
        file_path = Path(file_path)
        success, message = update_file(file_path)
        
        if success:
            updated += 1
            print(f"✓ {file_path.relative_to(BASE_DIR)} - {message}")
        elif "Error" in message:
            errors += 1
            print(f"✗ {file_path.relative_to(BASE_DIR)} - {message}")
        else:
            skipped += 1
            if "Already" in message or "Skipped" in message:
                # Only print skipped if verbose
                pass
            else:
                print(f"- {file_path.relative_to(BASE_DIR)} - {message}")
    
    print("-" * 80)
    print(f"Summary: {updated} updated, {skipped} skipped, {errors} errors")

if __name__ == "__main__":
    main()

File: scripts/update-navigation.js
#!/usr/bin/env node

/**
 * Script to update navigation menu across all HTML files
 * 
 * Usage: node scripts/update-navigation.js
 * 
 * This script finds all HTML files and updates the navigation menu
 * to ensure consistency across the site.
 */

const fs = require('fs');
const path = require('path');

// Configuration: Define navigation menu items in order
const GRINDING_POLISHING_MENU_ITEMS = [
    { url: '/metallographic-equipment/grinding-polishing/penta.html', label: 'Hand / Belt Grinders' },
    { url: '/metallographic-equipment/grinding-polishing/nano.html', label: 'Manual Grinder Polishers' },
    { url: '/metallographic-equipment/grinding-polishing/femto.html', label: 'Semi-Auto Grinder Polishers' },
    { url: '/metallographic-equipment/grinding-polishing/zeta.html', label: 'Automated Abrasive Dispenser' },
    { url: '/metallographic-equipment/grinding-polishing/atto.html', label: 'Controlled Removal Polishers' },
    { url: '/metallographic-equipment/grinding-polishing/giga.html', label: 'Vibratory Polishers' }
];

// Patterns to match navigation sections
const DESKTOP_NAV_PATTERN = /<ul class="sub-menu-second-level">\s*((?:<li><a href="[^"]*">[^<]*<\/a><\/li>\s*)*)<\/ul>/g;
const MOBILE_NAV_PATTERN = /<ul class="sub-menu second-level">\s*((?:<li><a href="[^"]*"><span>[^<]*<\/span><\/a><\/li>\s*)*)<\/ul>/g;

/**
 * Generate navigation HTML for desktop menu (just the list items, not the ul wrapper)
 */
function generateDesktopNav() {
    return GRINDING_POLISHING_MENU_ITEMS.map(item => 
        `                            <li><a href="${item.url}">${item.label}</a></li>`
    ).join('\n');
}

/**
 * Generate navigation HTML for mobile menu (just the list items, not the ul wrapper)
 */
function generateMobileNav() {
    return GRINDING_POLISHING_MENU_ITEMS.map(item => 
        `                            <li><a href="${item.url}"><span>${item.label}</span></a></li>`
    ).join('\n');
}

/**
 * Find and replace navigation section in HTML content
 */
function updateNavigation(content) {
    let updated = content;
    let changesMade = false;

    const newDesktopNav = generateDesktopNav();
    const newMobileNav = generateMobileNav();

    // Update desktop navigation - find the Grinding & Polishing section
    // Pattern: matches the ul tag and its content, replacing only the list items inside
    const desktopPattern = /(<li class="has-submenu">\s*<a href="\/metallographic-equipment\/grinding-polishing\.html">Grinding & Polishing<\/a>\s*<button[^>]*>[\s\S]*?<\/button>\s*<ul class="sub-menu-second-level">)\s*(?:<li><a href="\/metallographic-equipment\/grinding-polishing\/[^"]*">[^<]*<\/a><\/li>\s*)*(<\/ul>\s*<\/li>)/;
    const desktopMatch = updated.match(desktopPattern);
    if (desktopMatch) {
        updated = updated.replace(desktopPattern, `$1\n${newDesktopNav}\n                        $2`);
        changesMade = true;
    }

    // Update mobile navigation - find the Grinding & Polishing section
    // Pattern: matches the ul tag and its content, replacing only the list items inside
    const mobilePattern = /(<li>\s*<a href="\/metallographic-equipment\/grinding-polishing\.html">\s*<span>Grinding & Polishing[\s\S]*?<\/span>\s*<\/a>\s*<ul class="sub-menu second-level">)\s*(?:<li><a href="\/metallographic-equipment\/grinding-polishing\/[^"]*"><span>[^<]*<\/span><\/a><\/li>\s*)*(<\/ul>\s*<\/li>)/;
    const mobileMatch = updated.match(mobilePattern);
    if (mobileMatch) {
        updated = updated.replace(mobilePattern, `$1\n${newMobileNav}\n                        $2`);
        changesMade = true;
    }

    return { content: updated, changesMade };
}

/**
 * Recursively find all HTML files
 */
function findHtmlFiles(dir, fileList = []) {
    const files = fs.readdirSync(dir);
    
    files.forEach(file => {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);
        
        // Skip node_modules, .git, and other common directories
        if (file.startsWith('.') || file === 'node_modules' || file === '.git') {
            return;
        }
        
        if (stat.isDirectory()) {
            findHtmlFiles(filePath, fileList);
        } else if (file.endsWith('.html')) {
            fileList.push(filePath);
        }
    });
    
    return fileList;
}

/**
 * Main function
 */
function main() {
    const rootDir = path.join(__dirname, '..');
    const htmlFiles = findHtmlFiles(rootDir);
    
    console.log(`Found ${htmlFiles.length} HTML files`);
    
    let updatedCount = 0;
    let errorCount = 0;
    
    htmlFiles.forEach(filePath => {
        try {
            const content = fs.readFileSync(filePath, 'utf8');
            
            // Only process files that contain the Grinding & Polishing navigation
            if (content.includes('Grinding & Polishing') && 
                (content.includes('sub-menu-second-level') || content.includes('sub-menu second-level'))) {
                
                const { content: updatedContent, changesMade } = updateNavigation(content);
                
                if (changesMade) {
                    fs.writeFileSync(filePath, updatedContent, 'utf8');
                    console.log(`✓ Updated: ${path.relative(rootDir, filePath)}`);
                    updatedCount++;
                }
            }
        } catch (error) {
            console.error(`✗ Error processing ${filePath}:`, error.message);
            errorCount++;
        }
    });
    
    console.log(`\nSummary:`);
    console.log(`  Files updated: ${updatedCount}`);
    console.log(`  Errors: ${errorCount}`);
    console.log(`  Total files processed: ${htmlFiles.length}`);
}

// Run the script
if (require.main === module) {
    main();
}

module.exports = { updateNavigation, generateDesktopNav, generateMobileNav };

